test/parallel/test-async-wrap-pop-id-during-load.js

```diff
@@ -1,21 +1,22 @@
 'use strict';
 
-require('../common');
```

Comment:
Rather than move the module use to a later line, thus making this test violate our test formatting guidelines (and possibly make it inevitable that someone will just move it back some day soon), it would be better IMO to not call this file with spawnSync at all but instead to move the recursive async call stuff to a fixture. Is there a reason not to do that instead?

Reply:
+1 on this... the `require('../common')` should be at the top.

Reply:
Loading `common` is a immense overhead and when looking into that file there are only two reasons for it to always load currently:

1. On exit it checks if there are "leaked globals". I do not understand the reasoning for these checks. What is the point in checking for these? If anyone adds such a thing it should probably come up in a code review.
2. If the `NODE_TEST_WITH_ASYNC_HOOKS` environment variable is set, it will check for some things on exit. The reasoning for this was probably a rough smoke test. I personally believe we should not use it like this and instead just try to add a solid test base that would cover these things already. Do we know if this has ever uncovered some bugs?

Besides that `common` only exports functions for convenience.

So what I actually want to propose is to remove all `common` requires if they are not necessary and I would also like to split `common` into a couple small files that are specific to the individual use case. This will hopefully also improve the test runtime in general.

Reply:
The check for leaked globals should remain. It may be missed in code review and people should be notified when they run `make test` rather than having it come up during code review.

I'd be fine if everything else were moved to individual modules under `common` but that's a big change. But it would (probably) reduce the load times. I'm not really sure this working on shaving a second here and a second there from tests is an efficient approach anyway but I won't stop anyone who wants to do it. If it ends up making a big difference, great.

Reply:
But why would we manipulate globals in a test if there is no reason to do so?

Having a general check against leaked globals is absolutely fine but we could achieve that by just requiring all modules in a test and then checking for the known globals.

Reply:
Right now the check only tests for enumerable properties on global. When we add new globals, those are normally not enumerable. See https://github.com/nodejs/node/issues/20565 about this as well.

> That won't detect things if it only happens under certain conditions.
> If either the test or Node.js core accidentally creates a global, we want tests to fail.

Yes, that is correct. However, I highly doubt that we "accidentally" create a global. I do not remember any case in all the time that I worked on Node.js that someone wanted to add a global in a test or some code (if it was not explicitly about adding the global to the code). This could of course be the result of such test failures on local machines before someone opened a PR but I doubt that. Especially, since we only verify enumerable properties.

Please run:

```
node -e 'for (const val in global) console.log(val)'
```

The output should be:

```
global
process
Buffer
clearImmediate
clearInterval
clearTimeout
setImmediate
setInterval
setTimeout
__filename
exports
module
__dirname
require
```

As you can see, most of the "known globals" are not included at all (e.g. `global.Proxy`). Interestingly, `__filename` and `__dirname` show up as enumerable when calling it like this even though this is not the case when running our tests. I am not sure how this is set up.

We could now argue that the test should be fixed but to me it seems like we lived with a test that never really tested what we thought and we managed to deal with everything without.

So I still think it would be best to remove all those calls.

-----

That all aside: if we want to run some code in all our tests files all the time (no matter what code), should we not just rewrite the way how we call our tests and make sure the tests automatically load that code as well?

Reply:
@BridgeAR You make a reasonable case for removing the requirement of always loading `common`. I'm not sure I'm convinced--for example, if the current global leak detection approach does not work sufficiently well, might not the better approach be to fix it rather than to not require it? I'm asking rhetorically, as I can see arguments both ways on that one. If no one else objects to it, I won't either.

> That all aside: if we want to run some code in all our tests files all the time (no matter what code), should we not just rewrite the way how we call our tests and make sure the tests automatically load that code as well?

That had been suggested in the past and, IIRC, got a big -1 from people (I think @bnoordhuis was one of them) because it introduces more magic. In other words, tests will pass when called with `test.py` or whatever but fail if run directly from `node`. For me personally, I'm not sure that's really a problem. It might even be an improvement from our current situation if it changes us from "some tests fail if you don't use `test.py`" to "all tests fail if you don't use `test.py`". But I don't feel strongly about that.

@nodejs/testing

Reply:
Actually, thinking more about it, I'm not 100% sure the case that it doesn't work correctly is convincing. So what if it doesn't check non-enumerables? That just means we fail to detect it if someone adds a non-enumerable. First, that can be fixed. But second, in most cases, I'd imagine people accidentally leak enumerables. And those get detected. Or am I missing something?

Reply:
Yeah, it's only taken me four minutes, but I think I've returned to thinking that the global leak detection is valuable and should be in all tests.

Reply:
For me it still seems weird to have a halfway broken test that runs with each test file. Yes, it does detect enumerable entries but that should also be possible in almost all cases by just having a single test file that loads all modules. Of course the global could be added somewhere deep down but that is even more unlikely... I also do not know a compelling reason for not manipulating globals in tests. We normally do not do that but if it is done, it would only impact that specific test.  For me removing `common` is more about code hygiene than about runtime.

It feels like this is a legacy test that was implemented because 8 years ago the test suite was not that good and people pushed code directly to the repo instead of having code reviews. All that changed and as such this test became somewhat obsolete out of my perspective.

I do not want to spend much more energy into this but I really think we should start questioning some old decisions way more often.

Reply:
> I do not want to spend much more energy into this but I really think we should start questioning some old decisions way more often.

:100: to that. I think that's what we're doing here right now. We're just not (yet, at least) agreeing on what the correct answer to that questioning is. And you raise excellent points. I'm not convinced it outweighs the benefits, but it would be interesting to know what others think. AFACT, it's just you, me, @jasnell, and @TimothyGu that have weighed on in the specific issue.  Might be interesting to see what others think. This is a significant enough issue that I can see reconstituting the Testing WG if there are one or two other similar issues like this that need to be hashed out.